### PR TITLE
improve typing of DataArray and Dataset reductions

### PR DESCRIPTION
**Pull request by rhkleijn**
This PR makes the typing of reduction methods (`count`, `all`, `any`, `max`, `min`, `mean`, `prod`, `sum`, `std`, `var`, `median`) for `Dataset` and `DataArray` consistent with the typing of most other methods of `Dataset` and `DataArray` after the great recent improvements by @headtr1ck  in e.g. #6661.

### CODE REVIEW

**Comment by headtr1ck** [comment]
Noble effort, but I think we will have to wait until https://github.com/python/mypy/issues/12846 is solved, otherwise users will get false positives...

**Comment by rhkleijn** [comment]
Agreed, I just reached the same conclusion.

**Comment by headtr1ck** [comment]
Since https://github.com/python/mypy/issues/12846 seems to have been fixed, we could give this another go.

**Comment by rhkleijn** [comment]
Rebasing my branch required solving quite some merge conflicts, which was to be expected as this PR had run more than 400 commits behind. I adapted the PR to now use `Self` as you suggested.
It took a few iterations to satisfy mypy, but now CI is fully green.

File: xarray/core/computation.py [hunk 1]
```diff
@@ -1402,14 +1402,14 @@ def _cov_corr(
...
-        return cov  # type: ignore[return-value]
+        return cov  # type: ignore[return-value,unused-ignore]
 
     else:
         # compute std + corr
         da_a_std = da_a.std(dim=dim)
         da_b_std = da_b.std(dim=dim)
         corr = cov / (da_a_std * da_b_std)
-        return corr  # type: ignore[return-value]
+        return corr  # type: ignore[return-value,unused-ignore]
```
Comment (headtr1ck) [inline xarray/core/computation.py:1412]: Why do you need to add this? Can you not simply remove the ignore?

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -500,7 +500,7 @@ def __setitem__(self, key, value) -> None:
         self.dataset[dim_indexers] = value
 
 
-class Dataset(
+class Dataset(  # type: ignore[misc,unused-ignore]
```
Comment (headtr1ck) [inline xarray/core/dataset.py:503]: Why is this necessary? Can you post the error message that you get without?
Also, same question: why do you need to add unused-ignore?

**Review by headtr1ck — APPROVED** [review]
Thanks for updating this old PR.
Couple minor remarks.

Also, feel free to add an entry to what's new!

**Comment by headtr1ck** [comment]
Thanks for this PR!

**Comment by rhkleijn** [comment]
Thank you for taking care of the last few steps yourself and merging this PR! I just came back from vacation and just realized it would have been nicer if I had mentioned that I was going to be unresponsive for several weeks.
